### PR TITLE
[4.0] Favicon extensions support

### PR DESCRIPTION
**Pull request by 810**
Enabled by default (tracking bug) in:

    Chrome for desktop release 80
    Chrome for Android release 80

Consensus & Standardization
After a feature ships in Chrome, the values listed here are not guaranteed to be up to date.

    Firefox:    Shipped
    Edge:    No public signals
    Edge - chromium:   Shipped
    Safari:    No public signals
    Web Developers:    Positive

Pull Request for Issue # .

### CODE REVIEW

**Comment by brianteeman** [comment]
You need to update templateDetails.xml as well to list the new file

And I am guessing that `libraries\src\Document\OpensearchDocument.php ` also needs updating to support the svg

File: libraries/src/Document/OpensearchDocument.php [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 namespace Joomla\CMS\Document;
 
-\defined('JPATH_PLATFORM') or die;
+defined('JPATH_PLATFORM') or die;
```
Comment (Quy) [inline libraries/src/Document/OpensearchDocument.php]: Please revert. It is intentional when namespaced.

File: libraries/src/Document/OpensearchDocument.php [hunk 2]
```diff
@@ -128,7 +154,7 @@ public function render($cache = false, $params = array())
 	{
 		$xml = new \DOMDocument('1.0', 'utf-8');
 
-		if (\defined('JDEBUG') && JDEBUG)
+		if (defined('JDEBUG') && JDEBUG)
```
Comment (Quy) [inline libraries/src/Document/OpensearchDocument.php]: Please revert.

**Comment by HLeithner** [comment]
The code should be refactored and not only copy paste the ico code if we add svg we could png,gif and jpg too (not sure if it's in the specs but all browsers supports them)

**Comment by brianteeman** [comment]
As far as I understand the use of the mimetype is not recommended for current browsers. And anyway you have the wrong mimetypes

https://bitsofco.de/all-about-favicons-and-touch-icons/

**Comment by 810** [comment]
Safari and firefox is already supporting it. Chrome is supported on Feb 4. 

I believe there will be no joomla 4.0 beta before that date, so its save to add it now.

**Comment by brianteeman** [comment]
Thats not what I meant. The mimetype you have set in the code are not correct and its not recommended to use it anyway

**Comment by 810** [comment]
could you tell me which mimetype is wrong, for my knowledge they are correct.

from the demo page (source: chrome)
````
<link href="favicon.ico" rel="shortcut icon" type="image/vnd.microsoft.icon">
<link rel="icon" type="image/svg+xml" href="favicon.svg"/>
<link rel="alternate icon" type="image/png" href="favicon.png"/>
````

**Comment by brianteeman** [comment]
sorry looks like I was looking at it before you updated the types

**Comment by Quy** [comment]
Seems excessive on each page load to do 12 file checks/disk access.

**Comment by HLeithner** [comment]
> Seems excessive on each page load to do 12 file checks/disk access.

I wrote already a PR for this https://github.com/810/joomla-cms/pull/1

File: libraries/src/Document/HtmlDocument.php [hunk 3]
```diff
@@ -681,16 +682,25 @@ protected function _loadTemplate($directory, $filename)
...
-			if (file_exists($dir . $icon))
-			{
-				$path = str_replace(JPATH_BASE, '', $dir);
-				$path = str_replace('\\', '/', $path);
-				$this->addFavicon(Uri::base(true) . $path . $icon);
-				break;
+			$dir = \dir($path);
+
+			while (false !== ($entry = $d->read())) {
+				if (\array_key_exists($entry, $icons))
```
Comment (HLeithner) [inline libraries/src/Document/HtmlDocument.php]:
```suggestion
				if (\isset($icons[$entry]))
```

File: libraries/src/Document/OpensearchDocument.php [hunk 4]
```diff
@@ -70,46 +71,57 @@ public function __construct($options = array())
...
-			if (file_exists($dir . '/favicon.ico'))
-			{
-				$path = str_replace(JPATH_BASE, '', $dir);
-				$path = str_replace('\\', '/', $path);
-				$favicon = new OpensearchImage;
+			$dir = \dir($path);
 
-				if ($path == '')
+			while (false !== ($entry = $d->read())) {
+				if (\array_key_exists($entry, $icons))
```
Comment (HLeithner) [inline libraries/src/Document/OpensearchDocument.php]:
```suggestion
				if (\isset($icons[$entry]))
```

File: libraries/src/Document/OpensearchDocument.php [hunk 5]
```diff
@@ -70,46 +71,57 @@ public function __construct($options = array())
...
 		$update->template = Route::_(Uri::getInstance());
 		$this->addUrl($update);
 
 		// Add the favicon as the default image
 		// Try to find a favicon by checking the template and root folder
-		$app = Factory::getApplication();
-		$dirs = array(JPATH_THEMES . '/' . $app->getTemplate(), JPATH_BASE);
+		$app   = Factory::getApplication();
+		$dirs  = [JPATH_THEMES . '/' . $app->getTemplate(), JPATH_BASE];
+		$icons = ['/favicon.ico', '/favicon.gif', '/favicon.jpg', '/favicon.jpeg', '/favicon.png', '/favicon.svg'];
```
Comment (Quy) [inline libraries/src/Document/OpensearchDocument.php]:
```suggestion
```

File: libraries/src/Document/HtmlDocument.php [hunk 6]
```diff
@@ -20,6 +20,7 @@
 use Joomla\CMS\Utility\Utility;
 use Joomla\Database\ParameterType;
 use Joomla\Registry\Registry;
+use function array_key_exists;
```
Comment (Quy) [inline libraries/src/Document/HtmlDocument.php]:
```suggestion
```

File: libraries/src/Document/OpensearchDocument.php [hunk 7]
```diff
@@ -70,46 +71,59 @@ public function __construct($options = array())
...
+			while (false !== ($entry = $dir->read()))
+			{
+				if (isset($icons[$entry]))
 				{
-					if ($path[0] == '/')
+					$path    = str_replace(JPATH_BASE, '', $dir);
+					$path    = str_replace('\\', '/', $path);
+					$favicon = new OpensearchImage;
+
+					if ($path == '')
```
Comment (Quy) [inline libraries/src/Document/OpensearchDocument.php]:
```suggestion
					if ($path === '')
```

File: libraries/src/Document/OpensearchDocument.php [hunk 8]
```diff
@@ -70,46 +71,59 @@ public function __construct($options = array())
...
 					{
-						$path = substr($path, 1);
+						$favicon->data = Uri::base() . '/' . $entry;
 					}
+					else
+					{
+						if ($path[0] == '/')
```
Comment (Quy) [inline libraries/src/Document/OpensearchDocument.php]:
```suggestion
						if ($path[0] === '/')
```

File: libraries/src/Document/HtmlDocument.php [hunk 9]
```diff
@@ -681,16 +681,26 @@ protected function _loadTemplate($directory, $filename)
...
+			'favicon.png'  => 'image/png',
+			'favicon.svg'  => 'image/svg+xml',
+		];
+
+		foreach ([JPATH_BASE, $directory] as $path)
 		{
-			if (file_exists($dir . $icon))
+			$dir = \dir($path);
+
+			while (false !== ($entry = $dir->read()))
```
Comment (Quy) [inline libraries/src/Document/HtmlDocument.php]:
```suggestion
			while (($entry = $dir->read()) !== false)
```

File: libraries/src/Document/OpensearchDocument.php [hunk 10]
```diff
@@ -70,46 +71,59 @@ public function __construct($options = array())
...
-				$path = str_replace('\\', '/', $path);
-				$favicon = new OpensearchImage;
+			$dir = \dir($path);
 
-				if ($path == '')
-				{
-					$favicon->data = Uri::base() . 'favicon.ico';
-				}
-				else
+			while (false !== ($entry = $dir->read()))
```
Comment (Quy) [inline libraries/src/Document/OpensearchDocument.php]:
```suggestion
			while (($entry = $dir->read()) !== false)
```

File: libraries/src/Document/HtmlDocument.php (hunk 9 above)
Comment (810) [reply libraries/src/Document/HtmlDocument.php]: i tried some things, but im getting: 
````
Warning:
				dir(C:\wamp\www\mysql8\api\templates/system,C:\wamp\www\mysql8\api\templates/system): The system cannot
				find the path specified. (code: 3) in C:\wamp\www\mysql8\libraries\src\Document\HtmlDocument.php on line
				<i>695</i></th>
````

Comment (810) [reply libraries/src/Document/HtmlDocument.php]: when i do the str replacement then i get: Error: Call to a member function read() on bool: Could not match accept header

Comment (Quy) [reply libraries/src/Document/HtmlDocument.php]: Is this happening before my suggested change?

Comment (810) [reply libraries/src/Document/HtmlDocument.php]: the first message above is when i do your change.

And the second when i replace the windows dir to uri::base()

Comment (HLeithner) [reply libraries/src/Document/HtmlDocument.php]: Don't think so, it seams that the $directory variable has the wrong path "C:\wamp\www\mysql8\api\templates/system,C:\wamp\www\mysql8\api\templates/system" no Idea why

Comment (810) [reply libraries/src/Document/HtmlDocument.php]: i run that with the api test. with postman. You can see the error also on the drone tests beneath.

Comment (Quy) [reply libraries/src/Document/HtmlDocument.php]: I don't get this warning when testing using wamp.

Comment (810) [reply libraries/src/Document/HtmlDocument.php]: Frontend is now ok, no errors. Only the api tests are now failing

**Comment by 810** [comment]
still error on api test:
````

Body: <br />
--
206 | <b>Warning</b>:  dir(/tests/www/test-install/api/templates/system): failed to open dir: No such file or directory in <b>/tests/www/test-install/libraries/src/Document/HtmlDocument.php</b> on line <b>695</b><br />
207 | Error: Call to a member function read() on boolean: Could not match accept header
````

File: libraries/src/Document/OpensearchDocument.php [hunk 11]
```diff
@@ -70,46 +71,60 @@ public function __construct($options = array())
...
-				{
-					$favicon->data = Uri::base() . 'favicon.ico';
-				}
-				else
+			while (($entry = $dir->read()) !== false)
+			{
+				if (isset($icons[$entry]))
 				{
-					if ($path[0] == '/')
+					$path    = str_replace(JPATH_BASE, '', $dir);
```
Comment (Quy) [inline libraries/src/Document/OpensearchDocument.php]:
```suggestion
					$path    = \str_replace(JPATH_BASE, '', $dir);
```

File: libraries/src/Document/OpensearchDocument.php [hunk 12]
```diff
@@ -70,46 +71,60 @@ public function __construct($options = array())
...
+					$path    = str_replace('\\', '/', $path);
```
Comment (Quy) [inline libraries/src/Document/OpensearchDocument.php]:
```suggestion
					$path    = \str_replace('\\', '/', $path);
```

**Comment by brianteeman** [comment]
https://support.google.com/webmasters/answer/9290858

Your favicon should be a multiple of 48px square, for example: 48x48px, 96x96px, 144x144px and so on. SVG files, of course, do not have a specific size. Any valid favicon format is supported. Google will rescale your image to 16x16px for use in search results, so make sure that it looks good at that resolution. **Note: do not provide a 16x16px favicon.**

**Review by wilsonge — APPROVED** [review]

**Comment by 810** [comment]
RTC ?

**Comment by 810** [comment]
Ok chrome has released the new version with support, we can now set it to rtc

**Comment by Quy** [comment]
My concern is that the same file check is done with each page load. Would it be better to have a setting to select which format(s) have been installed?

**Comment by Quy** [comment]
@wilsonge Is my concern valid?

**Comment by wilsonge** [comment]
First thanks for the prod here :)

Where would you put the setting? I mean I agree that it's not ideal to loop through these things on each page load. but at the same time I'm not sure exactly where we could put the setting where it's logical.

If someone uploads an SVG for example - they put it in the active template directory. So logical is template component global config. But if they have multiple templates then it may not be the same file (or extension). But only the template can control it's own settings.

To me it's probably something we should cache. But I dunno if we have a mechanism within the cache driver to store specific pieces of information like this.

**Comment by coolcat-creations** [comment]
Hello everyone, I wonder if this plugin is useful for favicons in Joomla, if so I am happy to contribute it as a PR. 
(Before surely the code would need to be adapted to Joomla 4)

https://github.com/coolcat-creations/DBfavicon

    Adds favicons in different sizes + possibility to enable disable

    Adds apple-touch-icons in different sizes + possibility to enable disable

    Adds android-icons in different sizes
        Link behaviour
        Force screen orientation
        Customize Colour of task bar
        generates and adds manifest.json for android
        Customize Colour of task bar
        possibility to enable disable

    Adds Windows tiles
        Set up tile colour
        Adds browserconfig.xml file to root folder for mstiles
        possibility to enable disable
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27393">issues.joomla.org/tracker/joomla-cms/27393</a>.</sub>

**Comment by HLeithner** [comment]
> Hello everyone, I wonder if this plugin is useful for favicons in Joomla, if so I am happy to contribute it as a PR.
> (Before surely the code would need to be adapted to Joomla 4)
> 
> https://github.com/coolcat-creations/DBfavicon
> 
> ```
> Adds favicons in different sizes + possibility to enable disable
> 
> Adds apple-touch-icons in different sizes + possibility to enable disable
> 
> Adds android-icons in different sizes
>     Link behaviour
>     Force screen orientation
>     Customize Colour of task bar
>     generates and adds manifest.json for android
>     Customize Colour of task bar
>     possibility to enable disable
> 
> Adds Windows tiles
>     Set up tile colour
>     Adds browserconfig.xml file to root folder for mstiles
>     possibility to enable disable
> ```
> 
> This comment was created with the [J!Tracker Application](https://github.com/joomla/jissues) at [issues.joomla.org/tracker/joomla-cms/27393](https://issues.joomla.org/tracker/joomla-cms/27393).

I think a plugin would be the wrong approach for a native integration. But integrating this directly into HtmlDocument and in the template style part of joomla. This would allow every template to use it without knowing too much about the necessary settings.

One concern could be that this is a really fast moving target. My feeling is that ever 2 month there is a new property or resolution or other setting and two other get deprecated.

But I think this should a an own issue not not discussed as part of this pr.

**Comment by coolcat-creations** [comment]
> I think a plugin would be the wrong approach for a native integration. But integrating this directly into HtmlDocument and in the template style part of joomla. This would allow every template to use it without knowing too much about the necessary settings.
> 
> One concern could be that this is a really fast moving target. My feeling is that ever 2 month there is a new property or resolution or other setting and two other get deprecated.
> 
> But I think this should a an own issue not not discussed as part of this pr.

This code at least did not change since 4 Years. If you can help me how to implement it properly I am happy to contribute this to the core.

**Comment by brianteeman** [comment]
> This code at least did not change since 4 Years.

@HLeithner is talking about new formats like using svg and other formats being dropped
https://www.emergeinteractive.com/insights/detail/the-essentials-of-favicons/

**Comment by coolcat-creations** [comment]
> > This code at least did not change since 4 Years.
> 
> @HLeithner is talking about new formats like using svg and other formats being dropped
> https://www.emergeinteractive.com/insights/detail/the-essentials-of-favicons/

Yes and what I meant is that this plugin already has some more modern features than Joomla has now. Before implementing svg here we should know how to handle svg in Joomla at all.

**Comment by priyankaSahutekdi** [comment]
I have tested this item :white_check_mark: successfully on 173ea484e2614298b708200f40fdfa5e3842241f<br /><br />Verified in Chrome, Safari, Mozilla<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27393">issues.joomla.org/tracker/joomla-cms/27393</a>.</sub>

**Comment by punambaravkar** [comment]
I have tested this item :white_check_mark: successfully on 173ea484e2614298b708200f40fdfa5e3842241f<br /><br />Verified in Chrome, Safari, Mozilla<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27393">issues.joomla.org/tracker/joomla-cms/27393</a>.</sub>

**Comment by wilsonge** [comment]
@Quy any reason why you closed this one?

**Pull request closed without merging**